agent/agent.go

```diff
@@ -0,0 +1,628 @@
+	"fmt"
+	"hash/crc32"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"sync"
+	"time"
+
+	hb "github.com/m3db/m3em/generated/proto/heartbeat"
+	"github.com/m3db/m3em/generated/proto/m3em"
+	"github.com/m3db/m3em/os/exec"
+	"github.com/m3db/m3em/os/fs"
+
+	xerrors "github.com/m3db/m3x/errors"
+	"github.com/m3db/m3x/instrument"
+	xlog "github.com/m3db/m3x/log"
+	"github.com/uber-go/tally"
+	context "golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+)
+
+const (
+	defaultReportInterval       = time.Duration(5 * time.Second)
+	defaultHeartbeatConnTimeout = time.Minute
+	defaultTestCanaryPrefix     = "test-canary-file"
+)
+
+var (
+	errProcessMonitorNotDefined = fmt.Errorf("process monitor not defined")
+)
+
+type opAgent struct {
+	sync.RWMutex
+	opts           Options
+	logger         xlog.Logger
+	metrics        *opAgentMetrics
+	running        bool
+	token          string
+	executablePath string
+	configPath     string
+	processMonitor exec.ProcessMonitor
+	heartbeater    *opAgentHeartBeater
+}
+
+// New creates and retuns a new Operator Agent
+func New(
+	opts Options,
+) (Agent, error) {
+	if err := canaryWriteTest(opts.WorkingDirectory()); err != nil {
+		return nil, err
+	}
+
+	agent := &opAgent{
+		opts:    opts,
+		logger:  opts.InstrumentOptions().Logger(),
+		metrics: newAgentMetrics(opts.InstrumentOptions().MetricsScope()),
+	}
+	go agent.reportMetrics() // TODO(prateek): don't leak this
+	return agent, nil
+}
+
+func canaryWriteTest(dir string) error {
+	fi, err := os.Stat(dir)
+	if err != nil {
+		return fmt.Errorf("unable to stat directory, [ err = %v ]", err)
+	}
+	if !fi.IsDir() {
+		return fmt.Errorf("path is not a directory")
+	}
+
+	fd, err := ioutil.TempFile(dir, defaultTestCanaryPrefix)
+	if err != nil {
+		return fmt.Errorf("unable to create canary file, [ err = %v ]", err)
+	}
+	os.Remove(fd.Name())
+
+	return nil
+}
+
+func updateBoolGauge(b bool, m tally.Gauge) {
+	if b {
+		m.Update(1)
+	} else {
+		m.Update(0)
+	}
+}
+
+func (o *opAgent) reportMetrics() {
+	for {
+		running, exec, conf := o.state()
+		updateBoolGauge(running, o.metrics.running)
+		updateBoolGauge(exec != "", o.metrics.execTransferred)
+		updateBoolGauge(conf != "", o.metrics.confTransferred)
+		time.Sleep(defaultReportInterval)
+	}
+}
+
+func (o *opAgent) Running() bool {
+	o.RLock()
+	defer o.RUnlock()
+	return o.running
+}
+
+func (o *opAgent) state() (bool, string, string) {
+	o.RLock()
+	defer o.RUnlock()
+	return o.running, o.executablePath, o.configPath
+}
+
+func (o *opAgent) initFile(
+	fileType m3em.FileType,
+	filename string,
+	overwrite bool,
+) (*os.File, error) {
+	targetFile := path.Join(o.opts.WorkingDirectory(), filename)
+	if _, err := os.Stat(targetFile); os.IsExist(err) && !overwrite {
+		return nil, err
+	}
+
+	flags := os.O_CREATE | os.O_WRONLY
+	if overwrite {
+		flags = flags | os.O_TRUNC
+	}
+
+	perms := os.FileMode(0644)
+	if fileType == m3em.FileType_SERVICE_BINARY {
+		perms = os.FileMode(0755)
```

Comment:
Thought about this when writing it out but couldn't come up with a const name that wasn't more opaque. Stuck with this because _imo_ everyone should know what `0644` and `0755` mean.

Open to changing if you have a good const name.